tracer/src/Datadog.Trace/Debugger/Sink/BatchUploader.cs

```diff
@@ -0,0 +1,120 @@
+// <copyright file="BatchUploader.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Generic;
+using System.Text;
+using System.Threading.Tasks;
+using Datadog.Trace.Logging;
+
+namespace Datadog.Trace.Debugger.Sink
+{
+    internal class BatchUploader
+    {
+        private const int MaxSinglePayloadSize = 1 * 1024 * 1024;
+        private const int MaxTotalPayloadSize = 5 * 1024 * 1024;
+        private const int InitialBuilderSizeBytes = 10 * 1024;
+        internal const int InitialPayloadSizeBytes = 100 * 1024;
+
+        private const int InitialClosingBracketsLength = 2;
+        private const int SeparatorLength = 1;
+
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<BatchUploader>();
+
+        private readonly IBatchUploadApi _api;
+        private readonly StringBuilder _sb;
+
+        private byte[] _serializedPayloads = new byte[InitialPayloadSizeBytes];
+
+        private BatchUploader(IBatchUploadApi api)
+        {
+            _api = api;
+            _sb = new StringBuilder(InitialBuilderSizeBytes);
+        }
+
+        public static BatchUploader Create(IBatchUploadApi api)
+        {
+            return new BatchUploader(api);
+        }
+
+        public async Task Upload(IEnumerable<string> payloads)
+        {
+            try
+            {
+                foreach (var batch in GetBatches(payloads))
+                {
+                    await _api.SendBatchAsync(batch).ConfigureAwait(false);
+                }
+            }
+            catch (Exception e)
+            {
+                Log.Error(e, "Failed to upload batch");
+            }
+            finally
+            {
+                _sb.Clear();
+            }
+        }
+
+        private IEnumerable<ArraySegment<byte>> GetBatches(IEnumerable<string> payloads)
+        {
+            _sb.Clear();
+            _sb.Append('[');
+
+            var totalBatchSize = InitialClosingBracketsLength;
+            foreach (var payload in payloads)
+            {
+                var payloadSize = Encoding.UTF8.GetByteCount(payload);
```

Comment:
I don't have strong opinion on this.

What is the risk of sending big payloads?
What is the meaning of sending a lot of small payloads? 
What is the consequence of sending a big payload for client/backend?

@OmerRaviv?

Reply:
Yeah, this is quite unfortunate, as it is slow, and would quite often allocate on the LOH :(  But I think we should keep it for now, because of the following limitation in the Logs intake:

- The maximum limit for a batch of messages is 5MB.
- The maximum limit for a single message is 1MB.  The default behavior IIRC is that the message will still be stored, but the contents will simply be cropped out after the first 1MB.


This means we need to adapt to the following constraints and considerations.

1. We should avoid sending any message over 1MB, as we would not able to parse it and display it correctly in the variable tree in the Live Debugger UI.

2. We plan to add telemetry on how many messages were too large and lost due to this limitation in a follow-up PR.

3. We need to know whether adding this message to the existing batch will cause us to surpass the 5MB batch limit, in which case we need to start a new batch.

Long term, we hope to change things up to use a much more efficient serialization format for Live Debugger snapshots, but for the time being, I believe this is the best we can do.

Reply:
Yeah, I just realised we're using this in the DirectLogSubmission side too, so if we do find a way to fix it, would be nice to make it work there too 🙂 

I wonder if a quick fix would be to use `GetMaxByteCount()`, [as that is faster + provides an upper-bound](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Text/UTF8Encoding.cs#L758) on the serialized size. The risk there is it's quite conservative, so you could end up shipping smaller batches.

Either way, doesn't need doing in this PR, just a thought 🙂

Reply:
That's an interesting idea! Adding it to our list of follow-ups.